### PR TITLE
Add email verification & email request methods

### PR DESCRIPTION
**Pull request by nitishagar**
Connects-to: #855
Change-type: minor
HQ: https://github.com/balena-io/balena-io/pull/954
See: https://www.flowdock.com/app/rulemotion/resin-tech/threads/tNwuZRmXyhNdgKJZLq6-fckbOvC
Signed-off-by: Nitish Agarwal <1592163+nitishagar@users.noreply.github.com>
Depends-on: https://github.com/balena-io/balena-api/pull/3134

---
##### Contributor checklist

- [ ] Includes tests
- [x] Includes typings
- [x] Includes updated documentation
- [ ] Includes updated build output

### CODE REVIEW

File: lib/auth.ts
Comment (thgreasi) [inline lib/auth.ts:453]: How about:
```
* @fulfil {String|void} - session token or undefined in case that email verification is required.
```

File: lib/auth.ts
Comment (thgreasi) [inline lib/auth.ts:465]: How about having a single register call and inside the `then` have an `if` statement that does a `console.log('registrered)` or `console.log('Email verification is required')`?

**Review by thgreasi — COMMENTED** [review]
LGTM (Approved) but (reminder that) we need to wait for the API PR to get merged first, before merging this.

**Comment by JSReds** [comment]
@thgreasi @nitishagar Could we also update the User model with the properties added in the API model? IMHO It might be useful to have those properties. 

Referring to 
```
is_verified: boolean;
must_be_verified: boolean;
unverified_email: string | null;
...
```

**Comment by thgreasi** [comment]
@JSReds I think those are not whitelisted, so customers will not be able to use/know about these propeties.

**Comment by JSReds** [comment]
Could we whitelist them, I'm just asking since these might be useful data to make a correct redirect in case of an email verification UI

File: lib/auth.ts
Comment (thgreasi) [inline lib/auth.ts:557]: Let's use the public method:
```suggestion
		const id = await getUserId();
```

File: lib/auth.ts
Comment (thgreasi) [inline lib/auth.ts:556]: How about v ?
```suggestion
	async function resendVerificationEmail() {
```
I see a point though of this naming for pre-existing users that later try to verify their account, since "resend" doesn't make much sense if you have never been sent one before...

Comment (nitishagar) [reply lib/auth.ts:547]: Confused here - do you prefer the "resend" over "request"? From your comment doesn't look like it; but you change shows me that suggested.

File: lib/auth.ts
Comment (nitishagar) [reply lib/auth.ts]: Yes - post our discussion, I have updated sign-up to have a similar flow to log-in.

File: lib/auth.ts
Comment (dfunckt) [inline lib/auth.ts:463]: I don't see a reason for these changes. What is it you mean to convey to the user with the addition of the catch handler that merely rethrows the error?

File: lib/auth.ts
Comment (dfunckt) [inline lib/auth.ts:543]: Same comment here ^

Also, I'm either misreading the diff in the generated .md file above or maybe you forgot to regenerate that file after changes to this comment?

**Review by dfunckt — COMMENTED** [review]
Just a couple of comments related to doc-strings. Otherwise LGTM.

File: lib/auth.ts
Comment (nitishagar) [reply lib/auth.ts]: The reason I added them first was to cover both branches. But makes sense to not show the err' if we just show an example of throwing it again. Updated the PR.

File: lib/auth.ts
Comment (nitishagar) [reply lib/auth.ts]: Replied Above. And updated with required changes here.

File: DOCUMENTATION.md
Comment (thgreasi) [inline DOCUMENTATION.md:362]: Let's title the commit like
> Add email verification & email resend methods

File: lib/auth.ts
Comment (thgreasi) [inline lib/auth.ts:531]:
```suggestion
```

File: lib/auth.ts
Comment (thgreasi) [reply lib/auth.ts:542]: I'm in favor of my code change suggestion. The rest were just thoughts.

File: lib/auth.ts
Comment (thgreasi) [inline lib/auth.ts:462]: Why did we remove the second/callback-based example?
Let's revert the example changes here to keep the PR clean, since at the end we didn't change that method at all.
```suggestion
	 * }).then(function(token) {
	 * 	console.log(token);
	 * });
	 *
	 * @example
	 * balena.auth.register({
	 * 	email: 'johndoe@gmail.com',
	 * 	password: 'secret'
	 * }, function(error, token) {
	 * 	if (error) throw error;
	 * 	console.log(token);
	 * });
```
Feel free to open a separate PR though if you think this is anyway improving the examples.

File: lib/auth.ts
Comment (nitishagar) [reply lib/auth.ts:531]: Any reason we want to remove this? This would be only used after `login` and we need the following to work as well: https://github.com/balena-io/balena-sdk/pull/1077/files#diff-cd41b427b38a554f9a1a2d2e848e41ddc39bc7d68dba54c1398d7384d6aaf4f7R543

File: lib/auth.ts
Comment (nitishagar) [reply lib/auth.ts:542]: As you also mentioned lets keep the `requestVerificationEmail()` name given we are using this for pre-existing users as well.

File: lib/auth.ts
Comment (thgreasi) [reply lib/auth.ts:540]: I mean that I expect this to work even if you have logged in with an named user API key, which is possible for pre-existing users.
Not a big deal though I guess.

File: lib/auth.ts
Comment (thgreasi) [inline lib/auth.ts:542]:
```suggestion
```
Let's drop this, since we might drop the textual return value in the API soon.

File: lib/auth.ts
Comment (thgreasi) [inline lib/auth.ts:553]:
```suggestion
		await pine.patch({
```
Let's not return at all, since we might drop the textual return value in the API soon.

**Review by thgreasi — COMMENTED** [review]
Other than those LGTM

**Review by thgreasi — APPROVED** [review]

**Pull request merged**
